Revamp flash scripts

mbolivar opened the pull request:
This series:

- Fixes minor issues around flashing and DT that I noticed while poking around
- Re-works the flashing scripts to use command line arguments instead of environment variables
- Adds infrastructure to flash based on addresses in DT
- Illustrates DT-based flashing for pyocd and dfu-util (though a bit more is needed to make mcuboot app flashing "easy", which can't be added conveniently for reasons I'll explain if anyone asks)
- Adds some additional polish and convenience (e.g. print all available runners) to zephyr_flash_debug.py

It's meant for when the 1.11 merge window opens up.

@galak, @SebastianBoe, with this, it would be easy to support the features described in https://github.com/zephyrproject-rtos/zephyr/issues/4957. E.g. we could have each boards/common/foo.board.cmake (for foo = openocd, pyocd, dfu-util, etc.) add "foo" to a global property list of available flashers (or debuggers, depending on foo). The flash/CMakeLists.txt file could use that list to generate flash_foo, debug_foo, etc. targets.

However, the right way to flash and debug is under some discussion (e.g. https://github.com/zephyrproject-rtos/zephyr/issues/5153), so I'm not proposing those changes yet, pending decisions there.

Commit:
scripts: zephyr_flash_debug: simple improvements

Delete an unused import to make flake8 happy. Update the module
docstring.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: extract_dts_includes: fix potential undefined read

Only read 'status' when it has been assigned in compress_nodes().

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
dts: bindings: compatible properties are required

Any node which needs to generate defines from DTS must have a
"compatible" property, because the corresponding "constraint" is what
extract_dts_includes.py uses to match nodes with YAML files.

There are a few YAML files in the tree that list compatible as
"optional".  Fix them.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: refactor run() implementation

Have the subclasses implement a do_run() method instead, which run()
delegates to. This will make it possible to handle common
functionality in the superclass before runner-specific methods are
called. It is a prerequisite for tasks like loading the build time
configuration to add device tree awareness.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: pyocd: fix stale comment

This isn't just a flasher anymore.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: refactor flash/debug scripts to remove "shell"

The Python-based runners have replaced the old shell scripts. Refactor
the build system accordingly:

- FLASH_SCRIPT is now BOARD_FLASH_RUNNER
- DEBUG_SCRIPT is now BOARD_DEBUG_RUNNER

The values, rather than being the names of files, are now the names of
runners in scripts/support/runner. They are still short, descriptive
names like "openocd", "jlink", "em-starterkit", etc.

Adjust the zephyr_flash_debug.py call and runner internals
accordingly. Have each runner class report a name and the commands it
can handle. This lets us move some boilerplate from each do_run()
method into the common run() routine, and enables further improvements
in future patches.

The handles_command() method is temporary, and will be replaced by a
more general mechanism for describing runner capabilities in a
subsequent patch. The initial use case for extending this is to add
device tree awareness to the runners.

To try to avoid user confusion, abort the configuration if an
xxx_SCRIPT is defined.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: generalize commands to "capabilities"

Some configuration options or device tree nodes affect the way that
runners ought to behave, but there's no good way for them to report
whether they can handle them.

One motivating example is CONFIG_FLASH_LOAD_OFFSET, as influenced by
the zephyr,code-partition chosen node in the DT for architectures
where CONFIG_HAS_FLASH_LOAD_OFFSET=y.

If CONFIG_FLASH_LOAD_OFFSET is nonzero, the 'flash' command ought to
place the kernel at that address offset from the device flash's start
address. Runners don't support this right now, which should be
fixed. However, we don't want to mandate support for this feature,
since not all targets need it.

We need to let runners declare what their capabilities are. Make it so
by adding a RunnerCaps class to the runner core. This currently just
states which commands a runner can handle, but can be generalized to
implement the above use case.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: add build-time .conf parser

This parses generated_dts_board.conf and auto.conf, given a build
directory.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: use arguments, not environment vars

The various runners (flash/debug scripts) use environment variables to
take arguments. This is legacy behavior which is not desirable.

Use command line arguments instead.

Note: this leaves more general environment variables with publicly
      documented behavior in place for now, for compatibility, e.g.:

      ZEPHYR_FLASH_OVER_DFU, OPENSDA_FW, ESP_IDF_PATH, PYOCD_DAPARG

For example, when using dfu-util to flash arduino_101, instead of
setting DFUUTIL_PID, DFUUTIL_ALT, and DFUUTIL_IMG environment
variables, have the script invocation look like this:

  python3 .../zephyr_flash_debug.py dfu-util flash \
          [common arguments omitted] \
          --pid=8087:0aba --alt=x86_app \
          --img=.../build/zephyr/zephyr.bin

Make similar changes for other runners (openocd, etc.) and
targets (debug, debugserver).

To implement this in the scripts:

  - have the individual scripts/support/runner/some-runner.py files
    register their own command line arguments

  - teach them to construct instances from arguments, not the
    environment

  - have zephyr_flash_debug.py request runners to register command
    line argument parsers, and handle arguments

In the build system:

  - add a new board_runner_args() extension function that board.cmake
    files can use to add to the zephyr_flash_debug.py command line

  - adjust cmake/flash/CMakeLists.txt to invoke with arguments

  - add new helper include files for each runner (like
    boards/common/dfu-util.board.cmake, etc.), which add default
    options as needed and then add on overrides from
    board_runner_args() calls

  - update board.cmake files to use the new includes and extension

This implied some tweaking when using openocd to make the CMake string
escaping and unescaping work properly.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
cmake: extensions: add assert_not()

To be used in a later patch.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: add infrastructure for DT-based flashing

Add the necessary infrastructure to the runner core to support
computing flash addresses based on the devicetree. Specifically, add:

- a new RunnerCaps capability, flash_addr, which lets runners declare
  when they support flashing to an arbitrary address

- a common --dt-flash option to all runner command line parsers which
  support this capability, which lets users request flash addresses to
  be computed from device tree

- a ZephyrBinaryRunner helper method, get_flash_address(), which is
  the common code needed to compute a flash address from device
  tree (or return a default value if non-DT based flashing is
  requested). This relies on the BuildConfiguration parser introduced
  in an earlier patch.

Subsequent patches will use this functionality in individual runners.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
cmake: flash: support verbose output toggle

If CMAKE_VERBOSE_MAKEFILE is defined, make flashing and debugging more
verbose.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: zephyr_flash_debug: add 'runners' command

This just prints the available runner classes.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: let users copy/paste failed commands

Print a copy/pastable version of any command which fails, to help
debugging.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

dbkinder commented on boards/arm/v2m_beetle/doc/v2m_beetle.rst:
```diff
@@ -237,7 +237,7 @@ the :ref:`hello_world` application.
    :zephyr-app: samples/hello_world
    :board: v2m_beetle
-   :goals: build flash
+   :goals: build
```
I see you removed "flash" from the goals, but i don't see any replacement text for how to flash?  Perhaps your comment "the right way to flash and debug is under some discussion" means you don't have replacement text yet...

mbolivar replied:
Sorry to have been unclear. I tried to explain this change in the commit log:

```
This board doesn't support the 'flash' goal, which was mistakenly
included during the CMake conversion.
```

This is one of the "minor issues around flashing and DT that I noticed while poking around".

dbkinder reviewed (commented):
I presume your comment, "the right way to flash and debug is under some discussion" is why there's no replacement text for how to flash in the documentation change you made...

mbolivar commented:
@andrewboie, @carlescufi, @jhedberg and @MaureenHelm as code owners:

Feedback (especially testing on your platforms!) would be appreciated. @lpereira, you too -- I couldn't get esp32 to build following along with the documentation. Any confirmation that this is working for your boards would be great.

mbolivar commented:
> though a bit more is needed to make mcuboot app flashing "easy", which can't be added conveniently for reasons I'll explain if anyone asks

Let me add a bit more context here now that I've had some more time.

The basic issue is that mcuboot does not simply chain-load a binary at some flash device offset that's known to both the bootloader and the application linker script, but rather parses and validates a blob which includes digital signatures and other metadata located at that shared flash offset.

So for the zephyr scripts to flash something usable by mcuboot, they need additional command line arguments which can override what binary they flash to specify the output of the mcuboot postprocessor script (imgtool.py) instead of simply assuming zephyr.bin.

This is complicated because not all the flasher scripts use zephyr.bin. For example, the nrfjprog runner uses zephyr.hex.

So while it's a prerequisite to teach the flashers about arbitrary flash addresses, it's not the whole story.

carlescufi commented on cmake/flash/CMakeLists.txt:
```diff
@@ -1,48 +1,61 @@
+set(RUNNER_ARGS_COMMON
+  # Required:
+  "--board-dir=${BOARD_DIR}"
+  "--kernel-elf=${KERNEL_ELF_NAME}"
```
Why "kernel-*"? Wouldn't it be better to call it "zephyr-*"? This immediately confused me.

mbolivar replied:
The variable KERNEL_ELF_NAME is already present in CMake before this series. I'm using the same name in Python for consistency. Do you think we should change it?

carlescufi replied:
Yes, actually I think we should. Especially given that in the future we might decide to split the kernel into a different image for security or conformance reasons.

mbolivar replied:
@carlescufi: I looked into this further, and that change would be user-visible, since cmake/app/boilerplate.cmake gets all of these names from CONFIG_KERNEL_BIN_NAME. That would need to become CONFIG_ZEPHYR_BIN_NAME for consistency, which would break out of tree configs.

Given that and since this naming scheme is pre-existing, I'd like to regard this as out of scope for this series if OK with you.

carlescufi replied:
@mbolivar I understand. I will create a separate issue to deal with this, thanks for looking into it.

carlescufi commented on scripts/support/runner/core.py:
```diff
@@ -164,70 +194,189 @@ class ZephyrBinaryRunner(abc.ABC):
+        The standard (required) arguments are:
+
+        * --board-dir
+        * --kernel-elf, --kernel-hex, --kernel-bin
```
see previous comment

SebastianBoe commented on cmake/flash/CMakeLists.txt:
```diff
@@ -14,6 +14,15 @@ set(RUNNER_ARGS_COMMON
   )
 
+# Extra debugging, if requested.
+set(FLASH_DEBUG_VERBOSE OFF CACHE BOOL "If set, print verbose flash/debug output.")
```
The below is used elsewhere to solve the same problem.

    $<$<BOOL:${CMAKE_VERBOSE_MAKEFILE}>:--verbose>

I think it makes sense to re-use CMAKE_VERBOSE_MAKEFILE here,
because it is easier for a user if he only needs to learn the name of
one CMake variable to pass --verbose to underlying commands.

See zephyr/arch/x86/CMakeLists.txt

mbolivar replied:
Thanks; I wasn't aware of that. I'll fix it.

SebastianBoe reviewed (changes requested):
I like that configuration has moved from env vars to the cli for the flashers, and that different flashers can extend the CLI with their arguments.

But some of the changes are not obviously positive.

We want each layer to know as little as possible. Previously the flashing layer did not have access to the CONFIG_* values. The CMake layer however already knows what all CONFIG_* values are, so couldn't we avoid coupling the flashing layer to the Kconfig metadata by selectively passing on some CONFIG_* values from the CMake layer to the flashing layer?

I am afraid it could confuse a user if he sees two identical flashing commands that act differently because the flasher parses the .config file and acts differently.

EDIT: This is an example of where the runner is accessing CONFIG_* information directly.
https://github.com/zephyrproject-rtos/zephyr/pull/5232/files#diff-625597f38023505c804e54e3ab18ffcdR359

carlescufi commented:
First of all thanks @mbolivar for this PR. It addresses and reworks many of our concerns so far, especially the env variables dark magic and instead sets a solid foundation for command-line reusable tools for flashing and running.

@SebastianBoe I actually sort of disagree with your comments. I see no real drawback in accessing the Kconfig options from the flashers, because at the end of the day the configuration chosen by the user with Kconfig and DTS is something that needs to live through the whole process of building, flashing and running in order for the tools to have full knowledge of the metadata inside the .elf and .bin/.hex files. What would selective forwarding of certain Kconfig (and DTS?) options give us instead of having those available all through the process? I imagine for example that there will be new options in the future implemented as Python scripts and that will also require extensive information that can only be obtained from DTS and Kconfig.

SebastianBoe commented:
It might be possible / desirable to get rid of the CMake variables that are defined by board build scripts, and read by flasher build scripts.

We essentially have 2 interfaces to the flasher scripts, one that is a very thin
layer of CMake code that uses CMake variables. And then the one defined by
argparse. The CMake layer duplicates a lot of information with the argparse layer,
like whether an argument is optional or required.

Getting rid of the CMake layer would make everything more consistent and
easier to maintain.

See this as an example of how this could be done:
https://github.com/SebastianBoe/zephyr/commit/e8e9fb093e830b8d10c2b84389344d9c97b8450d

Now, not all information could be converted to this format, sometimes a board needs to set
a CMake variable so that it can be used in the build multiple places. But that is an exception,
rather than a rule.

@mbolivar : This architecture is inherited from KBuild, so you could dismiss this as out-of-scope if you want.

carlescufi commented:
> Getting rid of the CMake layer would make everything more consistent and
> easier to maintain.
> 
> See this as an example of how this could be done:
> SebastianBoe/zephyr@e8e9fb0

This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
+1 to that

SebastianBoe commented on scripts/support/runner/core.py:
```diff
@@ -142,15 +98,89 @@ def _parser_darwin(self, cmd):
+        kconfig = os.path.join(self.build_dir, 'zephyr', 'kconfig', 'include',
+                               'config')
+        files = [os.path.join(generated, 'generated_dts_board.conf'),
+                 os.path.join(kconfig, 'auto.conf')]
```
CMake parses values from ${DOTCONFIG} which is located somewhere else than auto.conf.

I don't know what the difference between auto.conf and ${DOTCONFIG} is. But I'm sure
we want CONFIG_* vars to be the same in the CMake namespace and in this namespace
so we should set them to the same file.

mbolivar replied:
Oh, OK, nice catch! That's a good point.

lpereira commented:
I'm taking a look re: ESP32.

mbolivar commented:
>> Getting rid of the CMake layer would make everything more consistent and
>> easier to maintain.
>> 
>> See this as an example of how this could be done:
>> SebastianBoe/zephyr@e8e9fb0
> This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
> +1 to that

I like this quite a bit.

I did this PR with CMake variables to minimize diff in the board files, but I'm happy to teach them about the python script directly (though this might open them up to a bit of churn, hopefully that's OK).

MaureenHelm commented on scripts/support/runner/core.py:
```diff
@@ -143,6 +143,17 @@ def _parser_darwin(self, cmd):
+    available in the commands field. This defaults to all three
+    commands.'''
+
+    def __init__(self, commands={'flash', 'debug', 'debugserver'}):
```
The commit message is a bit confusing since your changes don't actually check if a runner implements CONFIG_FLASH_LOAD_OFFSET. Are you suggesting that there will be a new command added for runners that implement it?

MaureenHelm replied:
Never mind, you answered this in a later commit

MaureenHelm reviewed (commented):
Thanks @mbolivar , this is really nice. I started thinking recently about how to add flashing capability to the jlink runner and bumped into one of the problems this PR works to solve. I also think it will help with creating a bootable image for mimxrt1050_evk, which has external flashes.

I'm afraid I'm in the middle of some PC issues though, so I won't be able to test on hardware for a few days.

mbolivar commented:
Updated.

- rebased (dropping two patches which made v1.10)
- added 9c016ed6b
- fixed s/auto.conf/.config/
- got rid of FLASH_DEBUG_VERBOSE

I need to test this some more tomorrow but want to provide an update for European TZs.

>>> Getting rid of the CMake layer would make everything more consistent and
>>> easier to maintain.
> 
>>> See this as an example of how this could be done:
>>> SebastianBoe/zephyr@e8e9fb0
>> This reduces the abstraction inside the CMake files by targetting directly the flash Python tool instead of storing state in intermediate, build-system only variables.
>> +1 to that
> I like this quite a bit.
> 
> I did this PR with CMake variables to minimize diff in the board files, but I'm happy to teach them about the python script directly (though this might open them up to a bit of churn, hopefully that's OK).

I tried to implement this, and now don't think it's a good idea. Here is why.

We agree that variables which board.cmake or the application CMakeLists.txt *must* set (i.e. mandatory variables with no default values, like NRF_FAMILY as pointed out by @SebastianBoe) can be safely converted to supply script arguments directly.

I think these cases should *not* be converted:

- User-settable variables (set at generation time or maybe "make" time, like board IDs / serial ports). Removing CMake variables means users cannot write "cmake -DVARIABLE=value".  Instead, they must edit CMake files, which is a worse experience, and hinders automation.

- Variables that board.cmake/application CMakeLists.txt may override/customize -- I don't mean set a flag, I mean change one non-boolean value to another. With CMake variables, it's easy for foo.board.cmake to test if FOO_SOME_VAR is unset, then supply default behavior for SOME_VAR. Adding options as strings directly to BOARD_RUNNER_ARGS_foo means having to do really ugly string processing instead, which is a worse implementation.

I analyzed the CMake variables which affect runners assuming these criteria:

https://gist.github.com/mbolivar/eb1d22872a4d6dc0b5022c08effe6d2d

9 can be converted, 27 cannot, and 2 I don't know. Moreover, in all but two cases (dfu-util and nrfjprog), the majority of each runner's variables cannot be converted.

And even cases like NRF_FAMILY and DFUUTIL_ALT are easier for board porters than adding options manually. It's easy for dfu-util.board.cmake to check if DFUUTIL_ALT is missing and supply a useful error message otherwise. That is easy for someone writing a board.cmake to understand. It's harder to understand what went wrong when a script they never heard of complains about a missing --pid option at runtime.

Given that most options can't be converted (so adding special cases makes the interface less consistent), and that there are advantages to keeping variables even in cases when they can, I argue that it's better to leave this as it is.

mbolivar commented:
> User-settable variables (set at generation time or maybe "make" time, like board IDs / serial ports). Removing CMake variables means users cannot write "cmake -DVARIABLE=value". Instead, they must edit CMake files, which is a worse experience, and hinders automation.

I guess I could support something like `RUNNER_foo_ARGS="--bar=quux --baz"` and append them on the global property, but that seems worse too.

SebastianBoe commented:
> I guess I could support something like RUNNER_foo_ARGS="--bar=quux --baz" and append them on the global property, but that seems worse too.

I was thinking that we could introduce a convention for targets so users can do

make <target>_args="--some-flag"

So it would be

make flash_args="--serial-nr=58076"

<target>_args are placed last on the command line and the script ensures that later arguments override earlier arguments, just like gcc usually does.

SebastianBoe commented:
> With CMake variables, it's easy for foo.board.cmake to test if FOO_SOME_VAR is unset, then supply default behavior for SOME_VAR. 

We can still support this use-case.

The app CMakeLists.txt can append flags.
The foo.board.cmake can prepend flags.

Conflicting flags are dealt with by the script e.g.

"-fno-exceptions -fexceptions" is interpreted as "-fexceptions"

The main benefit is perhaps that it is easier for users to discover options, as they are discoverable through the familiar "--help" flag.

They can do;

make flash_args=--help

and then

make flash_args=--openocd-gdb-port=2771

SebastianBoe commented:
A patchset like this is painful to maintain out of tree, so we can revisit this later if you are still sceptical @mbolivar .

mbolivar commented:
> We can still support this use-case.
> 
> The app CMakeLists.txt can append flags.
> The foo.board.cmake can prepend flags.
> 
> Conflicting flags are dealt with by the script e.g.
> 
> "-fno-exceptions -fexceptions" is interpreted as "-fexceptions"

I did think about "last one wins", but it doesn't work properly for cases like --dfuse-option, which may be given multiple times with cumulative effect. The fact that a case like that exists now means we still need the special-purpose string handling code in CMake.

> 
> The main benefit is perhaps that it is easier for users to discover options, as they are discoverable through the familiar "--help" flag.
> 
> They can do;
> 
> make flash_args=--help
> 
> and then
> 
> make flash_args=--openocd-gdb-port=2771

How can you do this? I thought we discussed that the only known way with cmake to pass a variable at runtime (in general, that would work with Ninja) was through the environment. Did I miss something? If so, that changes a lot of things for me...

> A patchset like this is painful to maintain out of tree, so we can revisit this later if you are still sceptical @mbolivar .

I'm OK to try to resolve this discussion before the merge; I hope we are getting close.

mbolivar commented:
> I did think about "last one wins", but it doesn't work properly for cases like --dfuse-option, which may be given multiple times with cumulative effect.

Let me think about getting rid of appendable options and otherwise verify that argparse can work in a "lst one wins" style. I think appending is useful, but being able to use the --help strings is indeed a nice thing.

SebastianBoe commented:
> How can you do this? I thought we discussed that the only known way with cmake to pass a variable at runtime (in general, that would work with Ninja) was through the environment. Did I miss something? If so, that changes a lot of things for me...

My mistake, this doesn't work with Ninja.

Perhaps the environment is the cleanest way to do this then. I think it would help usability if a user
could do:

$ flash_args="--help" ninja flash

Commit:
scripts: runner: pyocd: support DT-based flashing

Enable DT support in the pyOCD flasher, and turn it on in the build
system.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: pyocd: support pyocd-flashtool options

Allow the user to specify additional options for pyocd-flashtool. This
grants finer-grained control to individual applications, e.g. to
perform a chip erase instead of a sector erase when flashing.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: dfu-util: support DT-based flashing for DfuSe

Enable DT support in the dfu-util flasher when the target is a
DfuSe (DFU + ST extensions) device.

Untangling DfuSe-specific options (currently, the default is 'leave',
to immediately start running after the flashing is done) from the
actual address makes this cleaner, and sets up a subsequent patch to
let callers set DfuSe options.

It also lets us fix an unnecessary printline when flashing DfuSe
devices. There's no need to reset those, since the 'leave' modifier
starts execution immediately.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: dfu-util: add --dfuse-modifiers option

This gives callers finer-grained control over the flash. For example,
giving --dfuse-modifiers=force:mass-erase allows a chip-erase rather
than just erasing the sectors which need to be overwritten with the image.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
scripts: runner: add internal DEBUG to core

Add a debug-only global that prevents commands from running, and just
prints what would have been run.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

Commit:
cmake: extensions: support list arguments to set_ifndef()

I keep tripping over not being able to use set_ifndef to set a value
to a list, like this:

set_ifndef(VARIABLE value1 value2 value3...)

Allow that to work without changing its existing behavior when called
with one argument.

Signed-off-by: Marti Bolivar <marti@opensourcefoundries.com>

mbolivar commented:
A new update is available. I've tested on as many backends as I can (I've flashed arduino_101, 96b_carbon, nrf52_pca10040, frdm_k64f, disco_l475_iot1, and lightly debugged with those boards that support it). I used the (new in this version) DEBUG global in core.py to eyeball the command lines for other targets, and they seem sane.

The main difference from the previous version is that I've deleted almost all the CMake variable usage in boards/common/xxx.board.cmake, as discussed with @carlescufi and @SebastianBoe. The new extension function board_runner_args() should be used to add arguments directly instead in most cases.

You will see a few variables left (especially in openocd.board.cmake) where I thought they were helpful.

I've also added and removed a couple of less important patches.

> Perhaps the environment is the cleanest way to do this then. I think it would help usability if a user
> could do:
> 
> $ flash_args="--help" ninja flash

Sigh, I really want to get away from the environment, but I also hear what you're saying here. I'm hopeful if we keep working on this we'll find a solution that doesn't require the environment. Hopefully this series is a step towards it.

mbolivar commented:
> A new update is available. I've tested on as many backends as I can (I've flashed arduino_101, 96b_carbon, nrf52_pca10040, frdm_k64f, disco_l475_iot1, and lightly debugged with those boards that support it).

Scratch that -- I botched the rebase. Another version incoming

mbolivar commented:
> Scratch that -- I botched the rebase. Another version incoming

Nevermind, this should be fine -- I ran a last test in a shell with the wrong ZEPHYR_BASE. Silly environment variables :)
